baseplate/server/runtime_monitor.py

```diff
@@ -0,0 +1,168 @@
+                batch.gauge("gc.gen{}.{}".format(generation, name)).replace(value)
+
+
+class _GCTimingReporter(object):
+    def __init__(self):
+        try:
+            gc.callbacks
+        except AttributeError:
+            raise Exception("Python >=3.3 required")
+
+        gc.callbacks.append(self._on_gc_event)
+
+        self.gc_durations = []
+        self.current_gc_start = None
+
+    def _on_gc_event(self, phase, info):
+        if phase == "start":
+            self.current_gc_start = time.time()
+        elif phase == "stop":
+            if self.current_gc_start:
+                elapsed = time.time() - self.current_gc_start
+                self.current_gc_start = None
+                self.gc_durations.append(elapsed)
+
+    def report(self, batch):
+        gc_durations = self.gc_durations
+        self.gc_durations = []
+
+        for gc_duration in gc_durations:
+            batch.timer("gc.elapsed").send(gc_duration)
+
+
+def _report_runtime_metrics_periodically(metrics_client, reporters):
+    hostname = socket.gethostname()
+    pid = os.getpid()
+
+    while True:
+        now = time.time()
+        time_since_last_report = now % REPORT_INTERVAL_SECONDS
+        time_until_next_report = REPORT_INTERVAL_SECONDS - time_since_last_report
+        time.sleep(time_until_next_report)
+
+        try:
+            with metrics_client.batch() as batch:
+                batch.namespace += ".runtime.{}.PID{}".format(hostname, pid).encode()
+                for reporter in reporters:
+                    try:
+                        reporter.report(batch)
+                    except Exception as exc:
+                        logging.debug("Error generating server metrics: %s: %s",
+                                      reporter.__class__.__name__, exc)
+        except Exception as exc:
+            logging.debug("Error while sending server metrics: %s", exc)
+
+
+def start(server_config, application, pool):
+    if not hasattr(application, "baseplate") or not application.baseplate._metrics_client:
+        logging.info("No metrics client configured. Server metrics will not be sent.")
+        return
+
+    cfg = config.parse_config(server_config, {
+        "monitoring": {
+            "blocked_hub": config.Optional(config.Timespan, default=None),
+            "concurrency": config.Optional(config.Boolean, default=True),
+            "gc": {
+                "stats": config.Optional(config.Boolean, default=True),
+                "timing": config.Optional(config.Boolean, default=False),
+            },
+        },
+    })
+
+    reporters = []
+
+    if cfg.monitoring.concurrency:
+        reporters.append(_ConcurrencyReporter(pool))
+
+    if cfg.monitoring.blocked_hub is not None:
+        try:
+            reporters.append(_BlockedGeventHubReporter(cfg.monitoring.blocked_hub.total_seconds()))
+        except Exception as exc:
+            logging.info("monitoring.blocked_hub disabled: %s", exc)
```

Comment:
Maybe these messages should say `not supported by your current version of python/gevent` rather than `disabled`?

Reply:
It all sorta comes together into that:

```
monitoring.blocked_hub disabled: Gevent >=1.3 required
```

does that seem reasonable to you?

Reply:
ah, cool sgtm